Proposal for a Helm Deployment goal implementation (#15882)
pantsbuild/pants: docs/markdown/Helm/helm-deployments.md

```text
src/chart
src/docker
```

This should work any kind of Kubernetes resource that will result in Docker image being deployed into Kubernetes, like `Deployment`, `StatefulSet`, `ReplicaSet`, `CronJob`, etc. Please get in touch with us in case you find Pants was not capable to infer dependencies in any of your `helm_deployment` targets by either [opening a GitHub issue](https://github.com/pantsbuild/pants/issues/new/choose) or [joining our Slack](doc:getting-help).

> 📘 How the Docker image reference is calculated during deployment?
> 
> Pants' will rely on the behaviour of the `docker_image` target when it comes down to generate the final image reference. Since a given image may have more than one valid image reference, **Pants will try to use the first one that is not tagged as `latest`**, falling back into `latest` if none could be found.
> It's good practice to publish your Docker images using tags other than `latest` and Pants preferred behaviour is to choose those are this guarantees that the _version_ of the Docker image being deployed is the expected one.

Value override files
--------------------

> As seen in the initial examples of the usage of the `helm_deployment` target, it can be seen that some of the source value files used the word `override` in the file name.

Reviewer: Does this treat files with "override" in the name differently than the rest of the sources?

The StackStorm chart has a builds deployments for a lot of services so it's got a really long values file. So, I break my values files into a bunch of cross cuting concerns like resources.yaml, dns.yaml, annotations.yaml, and image.yaml. in some cases, the order they are passed to helm matters. So, how can I specify what order the values files are passed to helm?

Author: well, the usage of the `override` suffix may well be something that comes from my habits, met some other people that did the same, but defo up to debate.

I thought about leaving this out and just use the same order that is given in the `sources` field of the target however this can be problematic when using globs in that field if different computers compute a list of files in a different order.

The logic implemented would sort the value files making those that have the `override` word in their filename be last in the list passed to Helm. In a complex scenario like yours this is unlikely to be helpful if your files have a strict ordering.

Happy to remove this sorting feature and just use the ordering as listed in the target if we are sure that globing will always return files in the same order.

Reviewer: What if we did this (combing the override sort with the sources sort):
- preserve the order of entries in `sources`
- sort the file list within each glob, accounting for `override`:
    - alphanumerically sort for files without `override`:
    - alphanumerically sort for files with `override`
- explicitly listed files get sorted based on where they are in sources, even if they are included in the other globs

So in this example:
```suggestion
helm_deployment(name="dev", dependencies=["//src/chart"], sources=["common.yaml", "*.yaml", "dev/*.yaml", "dev/last.yaml"])
```
I would expect values files to be passed to `helm` in this order:
- common.yaml
- a.yaml
- z.yaml
- a-override.yaml
- z-override.yaml
- dev/a.yaml
- dev/z.yaml
- dev/a-override.yaml
- dev/z-override.yaml
- dev/last.yaml

Reviewer: We might be able to provide the overrides sorry with just globs too:
- preserve the order of entries in `sources`
- sort the file list within each glob:
    - alphanumerically sort for files
- explicitly listed files get sorted based on where they are in sources, even if they are included in the other globs

So in this example would achieve the same sort with this algorithm+sources as the previous algorithm+sources:
```python
helm_deployment(name="dev", dependencies=["//src/chart"], sources=["common.yaml", "*.yaml", "*override.yaml", "dev/*.yaml",  "dev/*-override.yaml", "dev/last.yaml"])
```

Author: got it, that would remove any magic meaning to the `override` suffix, which is not a bad thing.

It's a less trivial implementation though because it needs some sort of set hierarchy for each of the glob patterns, so a glob like `*.yaml` would ignore the files already included by the `*-override.yaml` pattern.

It may be possible to achieve using an instance of `PathGlobs` per pattern such that the other patterns are negated...

But at least it would be more intuitive from the user perspective.

Author: Implemented the new sorting approach and updated docs accordingly. The implementation can be found in the `_sort_value_file_names_for_evaluation` rule helper function

```
src/deployment/config_maps.yaml
src/deployment/services.yaml
src/deployment/dev/services.yaml
src/deployment/uat/services.yaml
```

```text src/deployment/BUILD
helm_deployment(name="dev", dependencies=["//src/chart"], sources=["*.yaml", "dev/*.yaml"])

helm_deployment(name="uat", dependencies=["//src/chart"], sources=["*.yaml", "uat/*.yaml"])
```

In this case, files that are in nested folders will also act as value override files.

We believe that this approach gives enough flexibility to organise your deployment sources as you better prefer while providing an intuitive way to Pants' users to ensure that the list of value file sources is in the right order every single time.

In addition to value files, you can also use inline values in your `helm_deployment` targets by means of the `values` field. All inlines values that are set this way will override any entry that may come from value files, whether the source file was an override file or not. The relationship between the three scopes can be imagined as follows:

```
Value files < Value override files < Inline values
```

Deploying
---------

> Continuing with the example in the previous section, we can deploy it into Kubernetes using the command `./pants experimental-deploy src/deployment`. This will go trigger the following steps:

Reviewer: ```suggestion
Continuing with the example in the previous section, we can deploy it into Kubernetes using the command `./pants experimental-deploy src/deployment`. This will trigger the following steps:
```

1. Analyse the dependencies of the given deployment.
2. Build and publish any first-party Docker image and Helm charts that are part of those dependencies.

> 3. Post-process the Kubernetes manifests generated by Helm by replacing all references to first-party Docker images by their real final registry destination.

Reviewer: Wow. You use post processing for that. Cool.

4. Initiate the deployment of the final Kubernetes resources resulting from the post-processing.

> The `experimental-deploy` goal also supports default Helm pass-through arguments that allow to change the deployment behaviour to be either atomic or a dry-run or even what is the Kubernetes config file and target context to be used in the deployment.

Reviewer: By "kubernetes config file" do you mean the helm values file?

Author: I meant the `kubeconfig` file (should be more specific about it) which can be specified to Helm using the `--kubeconfig` argument.

In CI it's quite probable that the file may live in a non-standard place (it does in our case) so it is useful to be able to pass that to the underlying Helm process.

> Please note that the list of valid pass-through arguments has been limited to those that not alter the reproducibility of the deployment (i.e. `--create-namespace` is not a valid pass-through argument). Those arguments will have equivalente fields in the `helm_deployment` target.

Reviewer: ```suggestion
Please note that the list of valid pass-through arguments has been limited to those that do not alter the reproducibility of the deployment (i.e. `--create-namespace` is not a valid pass-through argument). Those arguments will have equivalent fields in the `helm_deployment` target.
```

---
title: "Deployments"
slug: "helm-deployments"
hidden: false
createdAt: "2022-07-19T13:16:00.000Z"
updatedAt: "2022-07-19T13:16:00.000Z"
---
> 🚧 Helm deployment support is in alpha stage
> 

> > Pants has experimental support managing deployments via the `experimental-deploy` goal. Helm deployments provides with a basic implementation of this goal.

Reviewer 2: ```suggestion
> Pants has experimental support for managing deployments via the `experimental-deploy` goal. Helm deployments provides with a basic implementation of this goal.
```

> 
> Please share feedback for what you need to use Pants with your Helm deployments by either [opening a GitHub issue](https://github.com/pantsbuild/pants/issues/new/choose) or [joining our Slack](doc:getting-help)!

Motivation
----------

> Helm's ultimate purpose is to simplify the deployment of Kubernetes resources and help in making these reproducible. However is quite common to deploy the same software application into different kind of environments using slightly different configuration overrides.

Reviewer 2: ```suggestion
Helm's ultimate purpose is to simplify the deployment of Kubernetes resources and help in making these reproducible. However it is quite common to deploy the same software application into different kind of environments using slightly different configuration overrides.
```
